### PR TITLE
Assignment 4

### PR DESCRIPTION
**Pull request by jwoolbright23**

### CODE REVIEW

File: content/assignments/techjobs-persistent/_index.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+---
+title: "Assignment 4: Tech Jobs Persistent"
+date: 2023-01-03T10:42:56-06:00
+draft: false
+weight: 2
+originalAuthor: Sally Steuterman # to be set by page creator
+originalAuthorGitHub: gildedgardenia # to be set by page creator
+reviewer: Kimberly Horan # to be set by the page reviewer
+reviewerGitHub: codinglikeagirl42 # to be set by the page reviewer
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/_index.md:9]:
```suggestion
originalAuthor: John Woolbright # to be set by page creator
originalAuthorGitHub: jwoolbright23 # to be set by page creator
reviewer: Kimberly Horan # to be set by the page reviewer
reviewerGitHub: codinglikeagirl42 # to be set by the page reviewer
```

File: content/assignments/techjobs-persistent/part-four/index.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+## Refactor `Job.skills`
+
+1. Update your `Job` model class to fit its many-to-many relationship with skills.
+- `Job.skills` already exists. What needs to change and/or be added to map this relationship?
+- Be sure to check the whole class for any necessary type updates.
+
+## Updating `HomeController`, Again
+
+You next need to wire `HomeController` with the skills data in order to add skills objects to a new job.
+This will look almost precisely like what you have done for employer data above. Refer back to [that section](http://education.launchcode.org/java-web-dev-curriculum/assignments/techjobs-persistent/part-three/#updating-homecontroller) to inject the controller with skill data.
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/part-four/index.md:45]: Add a todo to fix link after merge.

File: content/assignments/techjobs-persistent/starter-code-review/index.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+lastEditor: # update any time edits are made after review
+lastEditorGitHub: # update any time edits are made after review
+lastMod: # UPDATE ANY TIME CHANGES ARE MADE
+---
+
+## Checkout and Review the Starter Code
+
+Set up a local copy of the project:
+
+1. In Canvas, **Graded Assignment #4: TechJobs (Persistent Edition)** contains a GitHub Classroom assignment invitation link and then set up the project in IntelliJ. Refer back to the GitHub Classroom instructions from [assignment0](https://education.launchcode.org/java-web-dev-curriculun/assignments/hello-world/index.html) for details.
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/starter-code-review/index.md:19]: Update wording to remove Github Classroom reference.

Add todo to update link

File: content/assignments/techjobs-persistent/starter-code-review/index.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+1. Launch the application (via the Gradle pane, *Tasks > Application > bootRun*) to make sure it starts up properly. Then shut it down.
+<!-- TODO: Link below will need to be updated once assignment 2 is in the book -->
+1. The tests for this assignment are set up the same way as for [assignment 2](https://education.launchcode.org/java-web-development/assignments/tech-jobs-oo.html#assignment-2-autograding).
+
+You won't be able to run your application or the tests yet. If you try to do so, you'll see a host of errors relating to the
+Spring Data annotations and classes. Some of these have already been used in the code, but the dependency that includes them has not yet been declared. That will be one of your tasks. You'll need to complete [Part1](https://education.launchcode.org/java-web-dev-curriculum/assignments/techjobs-persistent/part-one/#connect-a-database-to-a-spring-app) before you can
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/starter-code-review/index.md]:
```suggestion
Spring Data annotations and classes. Some of these have already been used in the code, but the dependency that includes them has not yet been declared. That will be one of your tasks. You'll need to complete [Part1]({{< relref "../part-one" >}}) before you can
```

File: content/assignments/techjobs-persistent/starter-code-review/index.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+run the application and view it in a browser.
+
+That said, it's a good idea to scan the classes and templates even before you're able to execute
+`bootRun`. Take a gander at the `Job` class. It will look somewhat similar to the model in
+[Tech Jobs (MVC Edition)](https://education.launchcode.org/java-web-development/assignments/tech-jobs-mvc.html#tech-jobs-mvc), with a few key differences.
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/starter-code-review/index.md:30]: Add todo to update link to new hugo book after merge

File: content/assignments/techjobs-persistent/starter-code-review/index.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+---
+
+## Checkout and Review the Starter Code
+
+Set up a local copy of the project:
+
+1. In Canvas, **Graded Assignment #4: TechJobs (Persistent Edition)** contains a GitHub Classroom assignment invitation link and then set up the project in IntelliJ. Refer back to the GitHub Classroom instructions from [assignment0](https://education.launchcode.org/java-web-dev-curriculun/assignments/hello-world/index.html) for details. 
+1. Launch the application (via the Gradle pane, *Tasks > Application > bootRun*) to make sure it starts up properly. Then shut it down.
+<!-- TODO: Link below will need to be updated once assignment 2 is in the book -->
+1. The tests for this assignment are set up the same way as for [assignment 2](https://education.launchcode.org/java-web-development/assignments/tech-jobs-oo.html#assignment-2-autograding).
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/starter-code-review/index.md]:
```suggestion
1. The tests for this assignment are set up the same way as for [assignment 2]().
```

File: content/assignments/techjobs-persistent/starter-code-review/index.md (hunk 5 above)
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/starter-code-review/index.md]:
```suggestion
<!-- TODO: Link below will need to be updated once assignment 3 is in the book -->
[Tech Jobs (MVC Edition)](), with a few key differences.
```

File: content/assignments/techjobs-persistent/starter-code-review/index.md (hunk 3 above)
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/starter-code-review/index.md]:
```suggestion
<!-- TODO: Link below will need to be updated once merged -->
1. In Canvas, **Graded Assignment #4: TechJobs (Persistent Edition)** contains a GitHub Classroom assignment invitation link and then set up the project in IntelliJ. Refer back to the GitHub Classroom instructions from [assignment0]() for details. 
```

File: content/assignments/techjobs-persistent/starter-code-review/index.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+Spring Data annotations and classes. Some of these have already been used in the code, but the dependency that includes them has not yet been declared. That will be one of your tasks. You'll need to complete [Part1](https://education.launchcode.org/java-web-dev-curriculum/assignments/techjobs-persistent/part-one/#connect-a-database-to-a-spring-app) before you can
+run the application and view it in a browser.
+
+That said, it's a good idea to scan the classes and templates even before you're able to execute
+`bootRun`. Take a gander at the `Job` class. It will look somewhat similar to the model in
+[Tech Jobs (MVC Edition)](https://education.launchcode.org/java-web-development/assignments/tech-jobs-mvc.html#tech-jobs-mvc), with a few key differences.
+
+You're no longer using a csv file to load job data, instead, we'll be creating new Job objects via a
+user form. The Job data will be stored in a MySQL database that you'll setup in [Part 1](http://education.launchcode.org/java-web-dev-curriculum/assignments/techjobs-persistent/part-one/#connect-a-database-to-a-spring-app) of this assignment.
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/starter-code-review/index.md]:
```suggestion
user form. The Job data will be stored in a MySQL database that you'll setup in [Part 1]({{< relref "../part-one/#connect-a-database-to-a-spring-app" >}}) of this assignment.
```

File: content/assignments/techjobs-persistent/starter-code-review/index.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+As you explore
+the starter code, you'll notice that the `JobField` abstract class is no longer present. Your task for
+[Part 2](https://education.launchcode.org/java-web-dev-curriculum/assignments/techjobs-persistent/part-two/#persisting-employers-and-skills) is to complete the work to persist some of the classes.
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/starter-code-review/index.md]:
```suggestion
[Part 2]({{< relref "../part-two/#persisting-employers-and-skills" >}}) is to complete the work to persist some of the classes.
```

File: content/assignments/techjobs-persistent/starter-code-review/index.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+You'll do this for `Employer` and `Skill` classes, as well as `Job`.
+
+The `Job` class will also look different from how you have last seen it. In Parts 3 and 4, you'll
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/starter-code-review/index.md]:
```suggestion
The `Job` class will also look different from how you have last seen it. In [Part 3]({{< relref "../part-three/#setting-up-a-one-to-many-relationship" >}}) and [Part 4]({{< relref "../part-four/#setting-up-a-many-to-many-relationship" >}}), you'll
```

File: content/assignments/techjobs-persistent/submission/index.md [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+## Sanity Check
+
+At this point, all autograding tests should be passing. To be sure, right-click on the `org.launchcode.techjobs.persistent` package in `src/test/java` and select *Run tests in...* If any test fails, evaluate the failure/error message and go back to fix your code.
+
+If a test in `TestCommentedTests` fails, this means that you failed to uncomment one or more tests in either `TestTaskTwo` or `TestTaskFour`. You will need to go back and uncomment the test(s) and make sure they pass.
+
+## How to Submit
+
+To turn in your assignment and get credit, follow the [Submission Instructions](https://education.launchcode.org/java-web-dev-curriculun/assignments/hello-world/index.html#submitting-your-work-on-canvas)
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/submission/index.md]: Add todo to update after merging.

Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/submission/index.md:23]:
```suggestion
To turn in your assignment and get credit, follow the [Submission Instructions]()
```

File: content/assignments/techjobs-persistent/part-one/index.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+Remember to double click on the schema name in the file tree to make it the default schema.
+{{% /notice %}}
+1. In the administration tab, create a new user, `techjobs` with the same settings as described in the lesson tutorial. This user should have the password `techjobs` as well.
+1. Update `build.gradle` with the necessary dependencies. At this point, you should be able to run the tests. Run the tests in `TestTaskOne` to verify your gradle dependencies.
+1. Update `src/resources/application.properties` with the correct info. This will include
+`spring.datasource.url` set to the address of your database connection, as well as the username and password for a user you have create Refer to the tip below for the other properties you must add to complete your database setup.
+
+{{% notice blue Tip "rocket" %}}
+You can double check your setup against what you've already done for [your coding events repo](https://education.launchcode.org/java-web-development/chapters/orm-part1/background.html#setup-orm-database). You can copy these property assignments from your coding events repo, only needing to change the database address and username/password values.
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/part-one/index.md]:
```suggestion
<!-- TODO: Link below will need to be updated once chapter 17 is merged in -->

You can double check your setup against what you've already done for [your coding events repo](). You can copy these property assignments from your coding events repo, only needing to change the database address and username/password values.
```

Comment (codinglikeagirl42) [reply content/assignments/techjobs-persistent/part-one/index.md]: Add todo to update when chapter 17 merged into the book.

File: content/assignments/techjobs-persistent/part-two/index.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+reviewer: Kimberly Horan # to be set by the page reviewer
+reviewerGitHub: codinglikeagirl42 # to be set by the page reviewer
+lastEditor: # update any time edits are made after review
+lastEditorGitHub: # update any time edits are made after review
+lastMod: # UPDATE ANY TIME CHANGES ARE MADE
+---
+
+## Persisting Employers and Skills
+
+You will need to have completed the [setup steps](https://education.launchcode.org/java-web-dev-curriculum/assignments/techjobs-persistent/part-one/#connect-a-database-to-a-spring-app) before starting this
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/part-two/index.md]:
```suggestion
You will need to have completed the [setup steps]({{< relref "../part-one/#connect-a-database-to-a-spring-app" >}}) before starting this
```

File: content/assignments/techjobs-persistent/part-two/index.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+In the last assignment, a `Job` object contained string fields for employer and core competency data. This employer
+and skill (formerly core competency) information about a particular job will now be stored in classes themselves.
+These items themselves will hold their own supplementary information.
+
+1. Open the `Employer` model class. In addition to the fields inherited from `AbstractEntity`, `Employer` should have a string field for `location`. Add the field for `location` with validation that ensures it is not empty and has a reasonable length. In addition, add public accessor methods to `Employer`.
+
+{{% notice blue Note "rocket" %}}
+For the purposes of this application, an employer can only have one location.
+{{% /notice %}}
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/part-two/index.md:42]:
```suggestion
   {{% notice blue Note "rocket" %}}
   For the purposes of this application, an employer can only have one location.
   {{% /notice %}}
```

File: content/assignments/techjobs-persistent/part-one/index.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+lastMod: # UPDATE ANY TIME CHANGES ARE MADE
+---
+
+## Connect a Database to a Spring App
+
+1. Start MySQL Workbench and create a new schema named `techjobs`.
+
+{{% notice blue Tip "rocket" %}}
+Remember to double click on the schema name in the file tree to make it the default schema.
+{{% /notice %}}
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/part-one/index.md:21]:
```suggestion
   {{% notice blue Tip "rocket" %}}
   Remember to double click on the schema name in the file tree to make it the default schema.
   {{% /notice %}}
```

Comment (codinglikeagirl42) [reply content/assignments/techjobs-persistent/part-one/index.md:21]: Formatting, tab in tip so ordered list formats correctly 1,2,3,4

File: content/assignments/techjobs-persistent/part-three/index.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+Then, make sure the data has been saved in your job table. You should see a column for `employer_id`, corresponding to the employer object selected for the new job.
+
+You have changed the architecture of your job table. You will still be able to add a new entry that has an`employer_id` column but you'll note that job still has the now defunct `employer` column. You can keep your database clean by removing the job table. It will be recreated when you run the application again.
+
+1. **SQL TASK:** In `queries.sql` under "Part 3", write the SQL statement to remove the job table.
+
+The *List* and *Search* functionality still isn't quite fixed so to view a job in the application, make a note
+of the job's id in the SQL table. Back in your browser, enter the path for `/view/{jobId}`.
+
+When everything works, move on to Part 4 below.
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/part-three/index.md]:
```suggestion
When everything works, move on to [ Part 4]({{< relref "../part-four" >}}).
```

File: content/assignments/techjobs-persistent/part-one/index.md [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+When your database is properly configured, you should have no compiler errors when starting the application. Execute `bootRun`
+and check the compiler output to make sure this is the case. If everything runs, you will be able to view your app
+locally in the browser at `http://localhost:8080` (unless of course you have changed the server port).
+
+1. In your MySQL workbench, open a new query tab to check your database connection.
+1. **SQL TASK:** At this point, you will have one table, `job`. In `queries.sql` under "Part 1", list the columns and their data types in the table as a SQL comment.
+
+Your running application still has limited functionality. You won't yet be able to add a job with the *Add Job* form. You also
+won't yet be able to view the list of jobs or search for jobs - but this is mostly because you have no job data. Move on to
+Part 2 below to start adding these functionalities.
```
Comment (codinglikeagirl42) [inline content/assignments/techjobs-persistent/part-one/index.md]:
```suggestion
[Part 2]({{< relref "../part-two" >}}) to start adding these functionalities.
```

File: content/assignments/techjobs-persistent/submission/index.md (hunk 10 above)
Comment (codinglikeagirl42) [reply content/assignments/techjobs-persistent/submission/index.md]:
```suggestion
<!-- TODO: Link below will need to be updated once merged -->

To turn in your assignment and get credit, follow the [Submission Instructions]()
```

**Review by codinglikeagirl42 — APPROVED** [review]
Review comments - accept suggestions and double check against trello checklist.

File: content/assignments/techjobs-persistent/part-two/index.md (hunk 13 above)
Comment (jwoolbright23) [reply content/assignments/techjobs-persistent/part-two/index.md:42]: This applies a strange font / box when tabbed in. Going to leave as is for now unless there is a fix I am unaware of.

Comment (jwoolbright23) [reply content/assignments/techjobs-persistent/part-two/index.md:42]: Ordered the numbers manually as a fix.

File: content/assignments/techjobs-persistent/part-one/index.md (hunk 14 above)
Comment (jwoolbright23) [reply content/assignments/techjobs-persistent/part-one/index.md:21]: Also ordered the numbers manually here to avoid funky formatting within note

**Pull request merged**